src/drivers/bootloaders/src/uavcan/main.c

```diff
@@ -753,85 +751,74 @@ static flash_error_t file_read_and_program(uint8_t fw_source_node_id,
-                 * !ValidReadReadResponse]:INDICATE_FW_UPDATE_INVALID_RESPONSE */
                 board_indicate(fw_update_invalid_response);
 
-                /* UAVCANBootloader_v0.3 #36: [(retries != 0 && timeout) ||
-                 * !ValidReadReadResponse]:1023.LogMessage.uavcan */
                 uavcan_tx_log_message(bootloader.node_id,
                                       UAVCAN_LOGMESSAGE_LEVEL_ERROR,
                                       UAVCAN_LOGMESSAGE_STAGE_PROGRAM,
                                       UAVCAN_LOGMESSAGE_RESULT_FAIL);
-                retries--;
+
             }
         }
 
+        /* Exhausted retries */
+
         if (can_status != CAN_OK)
         {
-            /* UAVCANBootloader_v0.3 #37: [(retries == 0 &&
-             * timeout]:INDICATE_FW_UPDATE_TIMEOUT */
             board_indicate(fw_update_timeout);
             break;
         }
 
         data = response.data;
-        write_length = response.data_length;
+
 
         /*
-         * If this is the last read (zero length) and there are bytes left to
-         * write, pad the firmware image out with zeros to ensure a full-word
-         * write. */
-        if (write_length == 0u && write_remainder_length > 0u)
-        {
-            write_length = 4u - write_remainder_length;
-            data[0] = data[1] = data[2] = 0u;
+         * STM32 flash addresses  must be word aligned
+         *
+         * If the packet is the last and an odd length
+         * add an 0xff to the end and bump the length
+         */
+
+        if (response.data_length & 1) {
+            data[response.data_length] = 0xff;
         }
 
-        /*
-         * If the length of a previous read was not a multiple of 4, we'll have a
-         * few bytes left over which need to be combined with the next write as
-         * all writes must be word-aligned and a whole number of words long. */
-        flash_status = FLASH_OK;
-        while (write_length && flash_status == FLASH_OK)
-        {
-            write_remainder[write_remainder_length++] = *(data++);
-            write_length--;
+        /* Save the first word off */
 
-            if (write_remainder_length == 4u)
-            {
-                if (bytes_written == 0u)
-                {
-                    /* UAVCANBootloader_v0.3 #30: SaveWord0 */
-                    memcpy(fw_word0, write_remainder, sizeof(uint32_t));
-                }
-                else
-                {
-                    flash_status =
-                        bl_flash_write_word((uint32_t)
-                                            (&bootloader.fw_image[bytes_written >> 2u]),
-                                            write_remainder);
-                }
+        if (request.offset == 0u)
+          {
+            bootloader.fw_word0.l = *(uint32_t *)data;
+            request.offset  += sizeof(uint32_t);
+            data += sizeof(uint32_t);
+            response.data_length -= sizeof(uint32_t);
```

Comment:
I don't see an easy way to change this. The file server can't guarantee any particular read size because it depends on remaining file length and behavior of the underlying `read()`.